### PR TITLE
Truncate UI service name if over 63 characters

### PR DESCRIPTION
**Pull request by jacobsalway**
## Purpose of this PR

Truncate the UI service name while appending a hash to ensure uniqueness if longer than 63 characters.

Close https://github.com/kubeflow/spark-operator/issues/2300

## Change Category

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [ ] Documentation update

## Checklist

- [X] I have conducted a self-review of my own code.
- [ ] I have updated documentation accordingly.
- [X] I have added tests that prove my changes are effective or that my feature works.
- [X] Existing unit tests pass locally with my changes.

### CODE REVIEW

File: pkg/util/sparkapplication.go [hunk 1]
```diff
@@ -156,7 +157,16 @@ func GetExecutorLocalVolumeMounts(app *v1beta2.SparkApplication) []corev1.Volume
...
+	if len(preferredName) <= 63 {
+		return preferredName
+	}
+
+	// Truncate the name and append a hash to ensure uniqueness while staying below the limit
+	hash := fmt.Sprintf("%x", md5.Sum([]byte(preferredName)))
+	return fmt.Sprintf("%s-%s-ui-svc", app.Name[:47], hash[:8])
 }
 
 func GetDefaultUIIngressName(app *v1beta2.SparkApplication) string {
```
Comment (ImpSy) [inline pkg/util/sparkapplication.go:172]: we should do the same for this function

File: pkg/util/sparkapplication.go [hunk 2]
```diff
@@ -156,7 +157,16 @@ func GetExecutorLocalVolumeMounts(app *v1beta2.SparkApplication) []corev1.Volume
...
-	return fmt.Sprintf("%s-ui-svc", app.Name)
+	preferredName := fmt.Sprintf("%s-ui-svc", app.Name)
+
+	// Service names are used as DNS labels, so must be 63 characters or shorter
+	if len(preferredName) <= 63 {
+		return preferredName
+	}
+
+	// Truncate the name and append a hash to ensure uniqueness while staying below the limit
+	hash := fmt.Sprintf("%x", md5.Sum([]byte(preferredName)))
```
Comment (ImpSy) [inline pkg/util/sparkapplication.go:168]: wdyt about reusing  the first 8 character from the submissionID (`app.Status.SubmissionID`) instead ?

File: pkg/util/sparkapplication.go (hunk 1 above)
Comment (jacobsalway) [reply pkg/util/sparkapplication.go:176]: Done

File: pkg/util/sparkapplication.go (hunk 2 above)
Comment (jacobsalway) [reply pkg/util/sparkapplication.go:168]: Not a bad idea either as that's a UUID. Truncating the name and appending its hash was the first approach that came to mind as I've seen it in other systems/software.

https://github.com/kubeflow/spark-operator/blob/master/internal/controller/sparkapplication/controller.go#L649

**Comment by jacobsalway** [comment]
Bumping since comments were addressed

**Review by ImpSy — APPROVED** [review]
Sorry for not reviewing this in a timely manner

**Comment by ImpSy** [comment]
@ChenYi015 can you approve this ?

**Review by ChenYi015 — APPROVED** [review]

**Pull request merged**